<span style = " font-size: 35px;"> **Tic Tac Toe Machine Learning Project**.</span>  
<span style = " font-size: 28px;"> **Introduction**.</span>  


This project enhances tic-tac-toe game by using machine learning techniques. It initially trains a decision tree classifier on game data to predict winners. Then, a Random Forest Classifier model is developed for improved prediction. Comparisons between the two models assess accuracy and evaluation metrics to determine performance.

<span style = " font-size: 30px;"> **Decision Tree**.</span> 

In [10]:
#importing necessary libraries
import pandas as pd
from sklearn.model_selection import train_test_split 
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix,classification_report, accuracy_score
from sklearn.preprocessing import LabelEncoder

In [11]:
 #Load dataset
filepath = 'tic-tac-toe-endgame.csv'
data = pd.read_csv(filepath)  
print(data.head())

  V1 V2 V3 V4 V5 V6 V7 V8 V9       V10
0  x  x  x  x  o  o  x  o  o  positive
1  x  x  x  x  o  o  o  x  o  positive
2  x  x  x  x  o  o  o  o  x  positive
3  x  x  x  x  o  o  o  b  b  positive
4  x  x  x  x  o  o  b  o  b  positive


__Lable Encoding__

In [12]:
 #convert label to numeric for this framework
le = LabelEncoder()
data['V10']= le.fit_transform(data['V10'])
data['V1']= le.fit_transform(data['V1'])
data['V2']= le.fit_transform(data['V2'])
data['V3']= le.fit_transform(data['V3'])
data['V4']= le.fit_transform(data['V4'])
data['V5']= le.fit_transform(data['V5'])
data['V6']= le.fit_transform(data['V6'])
data['V7']= le.fit_transform(data['V7'])
data['V8']= le.fit_transform(data['V8'])
data['V9']= le.fit_transform(data['V9'])
print(data.head())

   V1  V2  V3  V4  V5  V6  V7  V8  V9  V10
0   2   2   2   2   1   1   2   1   1    1
1   2   2   2   2   1   1   1   2   1    1
2   2   2   2   2   1   1   1   1   2    1
3   2   2   2   2   1   1   1   0   0    1
4   2   2   2   2   1   1   0   1   0    1


__Splitting into Train and Test Data sets__

In [13]:
#Preprocessing: extract independent-x and dependent-y variables        
y_data = data['V10']
X_data = data.drop('V10', axis =1)
X_train, X_test, y_train, y_test = train_test_split(X_data, y_data, test_size = 0.15, random_state = 0)
print(X_train.shape, ': Training data size')
print(X_test.shape, ': Test data size') 

(814, 9) : Training data size
(144, 9) : Test data size


__Training the Decision Tree Model__

In [14]:
DTclass = DecisionTreeClassifier(criterion='entropy', random_state=0)

# Train the Decision Tree Classifier Model
DTclass.fit(X_train, y_train) 

#The predict method is used to make predictions
pred_y = DTclass.predict(X_test)

__Evaluation of the trained Decision Tree Model__

In [15]:
C_matrix = confusion_matrix(y_test, pred_y)
print(classification_report(y_test, pred_y))

print("Confusion Matrix\n " , C_matrix)

print("\nAccuracy: = ", round(accuracy_score(y_test, pred_y),3))

              precision    recall  f1-score   support

           0       0.91      0.71      0.80        56
           1       0.84      0.95      0.89        88

    accuracy                           0.86       144
   macro avg       0.87      0.83      0.85       144
weighted avg       0.87      0.86      0.86       144

Confusion Matrix
  [[40 16]
 [ 4 84]]

Accuracy: =  0.861


__Analysing the classification report__

Precision: Measures the accuracy of positive predictions.
Recall: Measures the ability to identify all positive instances.
F1-score: Balances precision and recall.

The weighted average calculates the metrics by considering the support for each class, resulting in values of 0.87 for precision, 0.86 for recall, and 0.86 for the F1-score.

Overall, the model shows good performance, with higher precision, recall, and F1-score for class 1 compared to class 0, and an accuracy of 0.86.

<span style = " font-size: 30px;"> **Random Forest**.</span> 


In [16]:
#importing necessary libraries
import pandas as pd
from sklearn.model_selection import train_test_split 
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix,classification_report, accuracy_score
from sklearn.preprocessing import LabelEncoder

In [17]:
 #Load dataset  
filepath = 'tic-tac-toe-endgame.csv'
data = pd.read_csv(filepath)              
print(data.head()) 

  V1 V2 V3 V4 V5 V6 V7 V8 V9       V10
0  x  x  x  x  o  o  x  o  o  positive
1  x  x  x  x  o  o  o  x  o  positive
2  x  x  x  x  o  o  o  o  x  positive
3  x  x  x  x  o  o  o  b  b  positive
4  x  x  x  x  o  o  b  o  b  positive


__Lable Encoding__

In [18]:
 #convert label to numeric for this framework
le = LabelEncoder()
data['V10']= le.fit_transform(data['V10'])
data['V1']= le.fit_transform(data['V1'])
data['V2']= le.fit_transform(data['V2'])
data['V3']= le.fit_transform(data['V3'])
data['V4']= le.fit_transform(data['V4'])
data['V5']= le.fit_transform(data['V5'])
data['V6']= le.fit_transform(data['V6'])
data['V7']= le.fit_transform(data['V7'])
data['V8']= le.fit_transform(data['V8'])
data['V9']= le.fit_transform(data['V9'])
print(data.head())

   V1  V2  V3  V4  V5  V6  V7  V8  V9  V10
0   2   2   2   2   1   1   2   1   1    1
1   2   2   2   2   1   1   1   2   1    1
2   2   2   2   2   1   1   1   1   2    1
3   2   2   2   2   1   1   1   0   0    1
4   2   2   2   2   1   1   0   1   0    1


__Splitting into Train and Test Data sets__

In [19]:
#Preprocessing: extract independent-x and dependent-y variables        
y_data = data['V10']
X_data = data.drop('V10', axis =1)
X_train, X_test, y_train, y_test = train_test_split(X_data, y_data, test_size = 0.15, random_state = 0)
print(X_train.shape, ': Training data size')
print(X_test.shape, ': Test data size') 

(814, 9) : Training data size
(144, 9) : Test data size


__Training the Random Forest Model__

In [20]:
RFclass = RandomForestClassifier(n_estimators=20, bootstrap=True, criterion='entropy', random_state=0)

# Train the Random Forest Classifier Model
RFclass.fit(X_train, y_train) 

#The predict method is used to make predictions       
pred_y = RFclass.predict(X_test)

__Evaluation of the trained Random Forest Model__

In [21]:
Cmatrix = confusion_matrix(y_test, pred_y)

print(classification_report(y_test, pred_y))

print('Confusion Matrix: \n', Cmatrix)

print("\nAccuracy: = ", round(accuracy_score(y_test, pred_y),3))

              precision    recall  f1-score   support

           0       1.00      0.91      0.95        56
           1       0.95      1.00      0.97        88

    accuracy                           0.97       144
   macro avg       0.97      0.96      0.96       144
weighted avg       0.97      0.97      0.96       144

Confusion Matrix: 
 [[51  5]
 [ 0 88]]

Accuracy: =  0.965


__Analysing the classification report__

This classification report indicates that the model achieved high precision, recall, and F1-score for both 
classes, with 100% precision for class 0, 91% recall for class 0, 95% precision for class 1,
and 97% recall for class 1, resulting in an overall accuracy of 97%.

__Summary on the models used__

The Random Forest(RF) classifier model outperforms the Decision Tree(DT) classifier model significantly. 
It achieves higher precision, recall, and F1-score for both classes, with an accuracy of 97%, compared to the DT model's 86%. 
Additionally, the RF shows better balance between precision and recall, making it the preferred choice